### PR TITLE
Updates Byron-Shelley era specification implementation

### LINKED ISSUE #364: Implement PBFT rules and any missing dependency STSes in Haskell
The LaTeX specification of the Byron era has inference rules for the permissive BFT protocol (PBFT), but they haven't been implemented in the Haskell version of the specification. Implement PBFT inference rules (the PBFT STS) and any missing dependency STSes in the Haskell counterpart of the specification.

### PR DESCRIPTION
**Pull request by mdimjasevic**
This pull request updates the Haskell implementation of the Byron-Shelley era specification. This started as an implementation of the `PBFT` state transition system, but then that alone wasn't enough to get the whole system working.

I had to make changes to quite a few ledger STSes as well. In particular, can you @dnadales please take a look at all STSes that have `UPIState` as the type of their state? I'm not confident my changes are right.

Closes #364.

There are a few outstanding issues that I would need your help with:

* Clarke pointed out that `upAdptThd` and `t` are the same so use only one.
* Is the `dms` map needed in the `EPOCH` STS? As defined in "Figure 13: Epoch transition rules", I don't see it used anywhere.
* Is it `ProtocolConstants` or `PParams` that an [UTxOEnv id](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-38fe1250effe6d25f5cce1c3c2b6433dL82) should take? I assume the later and therefore I removed the [ProtocolConstants](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-38fe1250effe6d25f5cce1c3c2b6433dL85) type constructor.
* Ledger tests in `Ledger.Delegation.Examples` need to be fixed.
* I ~~commented out~~ removed an instance of [HasTrace CHAIN](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-3836797915c9b23257caef6f9996c602R107). @dnadales , do you see how to implement it now given the updated `CHAIN` STS?
* ~~The [bhToSign](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-5c0e6d4f8d7eeef6a96b94bafcc00054R125) function is undefined. How to define it?~~
* ~~The [bhHash](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-5c0e6d4f8d7eeef6a96b94bafcc00054R129) function is undefined too. How to define it?~~
* ~~The [bIsEBB](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-5c0e6d4f8d7eeef6a96b94bafcc00054R133) function is undefined. How to define it?~~
* ~~There are two variants of the `sEpoch` function. One is in [Chain.STS.Block.hs](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-5c0e6d4f8d7eeef6a96b94bafcc00054R142) and the other one is `sEpoch'` given in [Epoch.hs](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-450561489079056542e6bd3b22cf41f3R17).~~
* ~~The genesis UTxO is needed in the [BBODY](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-65f5ab0825dfa393387be1f28d2ff229R99) STS. Where and how is it defined in the concrete implementation? @ruhatch ?~~
* ~~A function that converts an `UTxO id` to `TxWits id` is needed, but [no such function is provided](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-65f5ab0825dfa393387be1f28d2ff229R100). How to implement one?~~
* ~~Does the `UPIEC` STS really take [no signal](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-450561489079056542e6bd3b22cf41f3R45)? If so, can it and the `PVBUMP` STS that it depends on still be implemented, given that it used to take an epoch as signal?~~
* ~~What about [k](https://github.com/input-output-hk/cardano-ledger-specs/pull/424/files#diff-76e54df1fec7e01bafbb2df1f06355c9R202)? Should it be 10 in the Byron-Shelley era?~~

### CODE REVIEW

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 1]
```diff
@@ -23,56 +26,97 @@ import Control.State.Transition
...
-              }
-        ds' <- trans @DELEG
-                     $ TRC (diEnv, ds, b ^. bBody . bDCerts)
-        return $! ds'
+        TRC ((ppsVal, e_n), (utxo, ds, us), b) <- judgmentContext
+        let bMax = ppsVal ^. maxBkSz
+        bSize b <= bMax ?! InvalidBlockSize
+        let bh = b ^. bHeader
+        hash (b ^. bBody ^. bUtxo)   == bh ^. bhUtxoHash ?! InvalidUtxoHash
+        hash (b ^. bBody ^. bDCerts) == bh ^. bhDlgHash  ?! InvalidUtxoHash
```
Comment (ruhatch) [inline byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: DelegationHash?

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 2]
```diff
@@ -23,56 +26,97 @@ import Control.State.Transition
...
+                    (fromList . keys . _dIStateDelegationMap) ds
+                , _dSEnvEpoch = e_n
+                , _dSEnvSlot = bh ^. bhSlot
+                }
+            )
+          , ds
+          , b ^. bBody ^. bDCerts
+          )
+        let
+          utxoGenesis = undefined :: UTxO TxId
```
Comment (ruhatch) [inline byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: You're right, this seems to be missing from `BBODY`, so maybe it needs to be added to the env there too.

In the implementation this comes from the genesis balances, in the genesis data. Genesis data should probably be part of the CHAIN environment too, so we should bubble it up through the rules.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 3]
```diff
@@ -23,56 +26,97 @@ import Control.State.Transition
...
+        -- TODO(md) 2019-04-30: The way things are now, a function of
+        -- type:
+        --
+        --   :: Ord id => UTxO id -> TxWits id
+        --
+        -- is needed such that b ^. bBody ^. bUtxo can be converted to
```
Comment (ruhatch) [inline byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: I think `bUTxO` has the wrong type - the UTxO payload should be a list of `TxWits`

Comment (ruhatch) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: In the LaTeX spec that is

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs [hunk 4]
```diff
@@ -1,23 +1,51 @@
...
+import Data.Bimap (Bimap)
 import Ledger.Core
+import Ledger.Update
+
+
+-- TODO(md): sEpoch used to take a map of type Map Epoch SlotCount,
+-- but with the latest revision of the Byron era chain spec (at least
+-- as of Apr 24, 2019) it does not anymore. For now I leave this
+-- function undefined.
+sEpoch' :: Slot -> Epoch
```
Comment (ruhatch) [inline byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs]: Ah, yeah, this is the right one - not sure how to deal with it though!

File: byron/chain/executable-spec/src/Cardano/Spec/Consensus/Block.hs [hunk 5]
```diff
@@ -21,11 +23,25 @@ class BlockHeader h where
...
+  bhDlgHash :: h -> Hash
+  -- | Update payload hash
+  bhUpdHash :: h -> Hash
 
 
 class BlockBody bb where
   -- | Delegation certificates.
   bbCerts :: bb -> [DCert]
+  -- | UTxO payload
+  bbUtxo :: bb -> UTxO TxId
```
Comment (ruhatch) [inline byron/chain/executable-spec/src/Cardano/Spec/Consensus/Block.hs]: Yeah, so this should probably be a list of `TxWits`

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs [hunk 6]
```diff
@@ -185,11 +195,15 @@ data DSEnv = DSEnv
...
   , _dSEnvSlot :: Slot
-  , _dSEnvStableAfter :: BlockCount
   } deriving (Show, Eq)
 
 makeFields ''DSEnv
 
+-- TODO(md) 2019-04-30: confirm this Byron-era value with the team
+-- | Chain stability parameter
+k :: BlockCount
+k = BlockCount 10
```
Comment (ruhatch) [inline byron/ledger/executable-spec/src/Ledger/Delegation.hs]: This is `2160` in Byron and multiplied by `10` to get the `epochSlots = 21600`

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 7]
```diff
@@ -12,6 +13,11 @@
 module Ledger.Update where
 
 import Control.Lens
+import qualified Crypto.Hash as Crypto
+import Data.AbstractSize (HasTypeReps)
+import qualified Data.Bimap as Bimap
```
Comment (ruhatch) [inline byron/ledger/executable-spec/src/Ledger/Update.hs]: Maybe add `import Data.Bimap (Bimap)` to clean up the type uses

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 1 above)
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Thanks for spotting this! I fixed it now.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs [hunk 8]
```diff
@@ -85,12 +117,31 @@ hashHeader bh = hashlazy . pack $
...
 
 -- | Compute the epoch for the given _absolute_ slot
-sEpoch :: Slot -> SlotCount -> Epoch
-sEpoch (Slot s) (SlotCount spe) = Epoch $ s `div` spe
+-- sEpoch :: Slot -> SlotCount -> Epoch
+-- sEpoch (Slot s) (SlotCount spe) = Epoch $ s `div` spe
+--
+-- TODO(md): to be implemented as defined in the LaTeX chain
+-- specification. The type matches that given in the specification.
+sEpoch :: Slot -> Map Epoch SlotCount -> Epoch
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs]: Right. I've put 21600 to be the fixed number of slots per epoch. Please shout if this is wrong for the Byron era!

File: byron/ledger/executable-spec/src/Ledger/Delegation.hs (hunk 6 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Delegation.hs]: Done!

File: byron/ledger/executable-spec/src/Ledger/Update.hs (hunk 7 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: Fixed.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 3 above)
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Fixed.

File: byron/chain/executable-spec/src/Cardano/Spec/Consensus/Block.hs (hunk 5 above)
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Consensus/Block.hs]: Fixed.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 2 above)
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: I added the genesis UTxO to the environments of `BBODY` and `CHAIN`.

File: byron/ledger/formal-spec/blockchain-interface.tex [hunk 9]
```diff
@@ -254,7 +255,8 @@ \subsection{Update-proposals interface}
       {\left(
         \begin{array}{l}
           s_n\\
-          \var{dms}
+          \var{dms}\\
+          \var{t}
```
Comment (nc6) [inline byron/ledger/formal-spec/blockchain-interface.tex]: This is a system level parameter, let's leave it as such. We're not likely to be changing it.

File: byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs [hunk 10]
```diff
@@ -55,17 +54,17 @@ deleg =
     ]
 
   , testGroup "Scheduling"
-    [ testCase "Example 0" $ checkTrace @SDELEG (DSEnv [gk 0, gk 1, gk 2] (e 8) (s 2) (bc 5)) $
+    [ testCase "Example 0" $ checkTrace @SDELEG (DSEnv [gk 0, gk 1, gk 2] (e 8) (s 2)) $
 
       pure (DSState [] [])
 
-    .- dc (gk 0) (k 10) (e 8) .-> DSState [(s 12, (gk 0, k 10))]
+    .- dc (gk 0) (k 10) (e 8) .-> DSState [(s 4322, (gk 0, k 10))]
```
Comment (nc6) [inline byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs]: What's the motivation behind this change? 4322 seems like a very strange number to hard-code

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 11]
```diff
@@ -23,56 +26,88 @@ import Control.State.Transition
...
-  type Environment BBODY
-    = ( Epoch
-      , Slot
-      , PParams
-      , Set VKeyGenesis
-      )
+  type Environment BBODY =
+    ( PParams
+    , Epoch
+    , UTxO TxId
```
[the remark this answers is not in the document]
Comment (nc6) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Sorry, seen that the spec has been updated.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 12]
```diff
@@ -23,56 +26,88 @@ import Control.State.Transition
...
+    , BlockCount -- ^ The t parameter expressed in terms of the number
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: The `t` parameter is passed to `BUPI`'s environment.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Bupi.hs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+import Control.State.Transition
+  ( Embed
+  , Environment
+  , PredicateFailure
+  , STS
+  , Signal
+  , State
+ , TRC(TRC)
+--  , (?!)
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Bupi.hs]: Right, not needed. I'm removing it in the upcoming commit.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs [hunk 14]
```diff
@@ -1,23 +1,48 @@
...
   type Signal EPOCH = Slot
-  data PredicateFailure EPOCH = X
-    deriving (Eq, Show)
+  data PredicateFailure EPOCH =
+    NotAfterCurrentEpoch Epoch Epoch deriving (Eq, Show)
+
   initialRules = []
+
+  -- In the LaTeX counterpart there is a transition rule that does not
+  -- change the state at all so it is left out here.
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs]: I see no difference between having no plausible rule to fire and a rule that fires, but does not make any progress and instead creates an infinite loop. Would you mind explaining the difference?

File: byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs (hunk 10 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs]: It's not any different than 12. I could use someone's help here as I don't fully understand the test.

Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs]: The tests are just testing that given the initial states, and the signals, we get the corresponding final states. These are just random numbers, but I think @nc6 's point was that if we start with some arbitrary random numbers, there shouldn't be a reason for changing them to some other numbers. 

Also if you have change a value in the signal you'd have to update the other values that come after that signal.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs [hunk 15]
```diff
@@ -85,12 +116,25 @@ hashHeader bh = hashlazy . pack $
 -- TODO: we might want to serialize this properly, without using show...
 
 -- | Computes the hash of the header.
-bhToSign :: BlockHeader -> Hash
-bhToSign = hashHeader
+--
+-- TODO(md): This is not defined in the LaTeX specification and it is
+-- not clear what type it should have. The current signature is
+-- probably wrong, but it is here just to get type-checking working
+-- w.r.t. using the verify function in a rule for PBFT.
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs]: Done.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs [hunk 16]
```diff
@@ -85,12 +116,25 @@ hashHeader bh = hashlazy . pack $
...
+bhToSign :: BlockHeader -> VKey
+bhToSign = undefined
+
+-- This is used in PBFT and CHAIN STSes
+bhHash :: BlockHeader -> Hash
+bhHash = undefined
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Block.hs]: Done.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    | PrevHashNotMatching Hash Hash
+    | InvalidHeaderSignature VKey (Sig VKey)
+    | SigCountFailure (PredicateFailure SIGCNT)
+    deriving (Eq, Show)
+
+  initialRules = []
+
+  transitionRules =
+    [ do
+        -- TODO(md) [2019-04-24]: see if all checks from the prose are
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs]: Yes, that's the one. What's your take? Should the prose be changed or is it fine the way it is?

File: byron/ledger/formal-spec/blockchain-interface.tex (hunk 9 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/blockchain-interface.tex]: Are you saying it shouldn't be in any STS'es environment, for example the way `k` is not in any environment any more?

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Epoch.hs]: After thinking about this for a while, I believe I see how an STS is meant to work. With that in mind, I'll implement these rules that I left out.

File: byron/ledger/executable-spec/src/Ledger/UTxO.hs [hunk 18]
```diff
@@ -184,6 +206,36 @@ instance Ord id => STS (UTXOW id) where
...
+  type State (UTXOWS id) = UTxO id
+  type Signal (UTXOWS id) = [TxWits id]
+  type Environment (UTXOWS id) = UTxOEnv id
+  data PredicateFailure (UTXOWS id)
+    = UtxowFailure (PredicateFailure (UTXOW id))
+    deriving (Eq, Show)
+
+  -- There is a rule in the LaTeX document for this STS that makes no
+  -- change to the state without any preconditions, hence this rule is
+  -- not imeplemtened as it is a no-op.
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/UTxO.hs]: Fixed it.

File: byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs (hunk 10 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs]: I'm not sure I'm following. What needs to be done with this test (if anything)?

Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Delegation/Examples.hs]: Nothing. Just revert back 4332 to 12, or change the ensuing states accordingly.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nc6) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: I'm saying this doesn't explain anything about what this parameter _is_

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs (hunk 17 above)
[the remark this answers is not in the document]
Comment (nc6) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs]: Could change the text, but it's not high priority. Could you create an issue for this?

File: byron/ledger/formal-spec/blockchain-interface.tex (hunk 9 above)
Comment (nc6) [reply byron/ledger/formal-spec/blockchain-interface.tex]: Yes

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: It is a threshold expressed as a block count needed by the `UPEND` STS, which is for protocol version endorsement. In the ledger spec it corresponds to `q * ngk` in `UPEND`'s environment. Does that make sense?

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs (hunk 17 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Pbft.hs]: Sure! Here it is: https://github.com/input-output-hk/cardano-ledger-specs/issues/458

File: byron/ledger/formal-spec/blockchain-interface.tex (hunk 9 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/blockchain-interface.tex]: There is `upAdptThd`, which is available in protocol parameters. I'm not completely sure what's happening here, i.e., how it is related to `t`. Can you help here?

File: byron/ledger/executable-spec/src/Ledger/UTxO.hs [hunk 19]
```diff
@@ -79,11 +89,13 @@ instance Show id => BA.ByteArrayAccess (Tx id) where
...
-                          , pps   :: ProtocolConstants id
+                          , pps   :: PParams
                           }
 
-newtype ProtocolConstants id = ProtocolConstants
-  { pcMinFee :: Tx id -> Value }
+-- newtype ProtocolConstants id = ProtocolConstants
+--   { pcMinFee :: Tx id -> Value }
+pcMinFee :: PParams -> Tx id -> Value
+pcMinFee = undefined
```
Comment (nc6) [inline byron/ledger/executable-spec/src/Ledger/UTxO.hs]: Needs to be fixed before merging

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 20]
```diff
@@ -673,6 +721,37 @@ instance STS UPIVOTE where
...
+  type State UPIVOTES = UPIState
+  type Signal UPIVOTES = [Vote]
+
+  data PredicateFailure UPIVOTES
+    = UpivoteFailure (PredicateFailure UPIVOTE)
+    deriving (Eq, Show)
+
+  -- There is a rule in the LaTeX document for this STS that makes no
+  -- change to the state without any preconditions, hence this rule is
+  -- not imeplemtened as it is a no-op.
```
Comment (nc6) [inline byron/ledger/executable-spec/src/Ledger/Update.hs]: OK, looking in more detail, it seems that what you actually mean is that you've folded the two rules into one. As it is, this comment is deeply misleading, and somewhat harmful since no-op rules _must_ be implemented to preserve the system being able to move forward.

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 21]
```diff
@@ -723,17 +799,23 @@ instance Embed UPEND UPIEND where
...
+    -- )
+    ( Core.Slot
     , [(Core.Slot, (ProtVer, PParams))]
     )
-  type Signal PVBUMP = Core.Epoch
+  type State PVBUMP =
+    -- ( Core.Epoch
+    -- , (ProtVer, PParams)
+    -- , [(Core.Slot, (ProtVer, PParams))]
+    -- )
```
Comment (nc6) [inline byron/ledger/executable-spec/src/Ledger/Update.hs]: Please tidy these before merging

File: byron/ledger/executable-spec/src/Ledger/Update.hs [hunk 22]
```diff
@@ -742,60 +824,92 @@ instance STS PVBUMP where
...
+  --       TRC (_, (ep, (pv, pps), fads), en) <- judgmentContext
+  --       en <= ep ?! NewEpoch
+  --       return (ep, (pv, pps), fads)
+  --   , do
+  --       TRC ((k, sn), (ep, (pv, pps), fads), en) <- judgmentContext
+  --       ep < en ?! OldEpoch
+  --       return $ case (partition (\(s, _) -> s <= sn `Core.minusSlot` liveAfter k) fads) of
+  --         ([], _) -> (ep, (pv, pps), fads)
+  --         ((_, (pvc, ppsc)):_, rest) -> (en, (pvc, ppsc), rest)
+  --   ]
```
Comment (nc6) [inline byron/ledger/executable-spec/src/Ledger/Update.hs]: Remove

File: byron/ledger/formal-spec/blockchain-interface.tex (hunk 9 above)
Comment (nc6) [reply byron/ledger/formal-spec/blockchain-interface.tex]: Removing the parameter in the separate PR is ok, but the incorrect changes to this document should be addressed before this PR is merged. There's no sense in merging something known to be incorrect and then having a separate PR to revert it.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs [hunk 23]
```diff
@@ -23,56 +26,87 @@ import Control.State.Transition
...
-    = ( Epoch
-      , Slot
-      , PParams
-      , Set VKeyGenesis
-      )
+  type Environment BBODY =
+    ( PParams
+    , Epoch
+    , UTxO TxId
+    , BlockCount
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Haven't we agreed to address remaining issues in separate issues and PRs? I'm simply addressing issues that the CI has ran into.

File: byron/ledger/executable-spec/src/Ledger/Update.hs (hunk 20 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: You're right and this is an outdated comment, as you noticed by comparing it with the implementation. I'm removing the comment...

File: byron/ledger/executable-spec/src/Ledger/Update.hs (hunk 22 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: Done.

File: byron/ledger/executable-spec/src/Ledger/Update.hs (hunk 21 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/src/Ledger/Update.hs]: Done.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 23 above)
[the remark this answers is not in the document]
Comment (nc6) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Some things it might make sense to defer, but not where incorrect things have been introduced in this PR.

File: byron/ledger/formal-spec/blockchain-interface.tex (hunk 9 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/blockchain-interface.tex]: Just fixed this.

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs (hunk 23 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/BBody.hs]: Removed from the environment!

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Chain.hs [hunk 24]
```diff
@@ -7,206 +7,106 @@
...
 
 data CHAIN
 
 instance STS CHAIN where
   type Environment CHAIN =
     ( Slot            -- Current slot
-    , Set VKeyGenesis -- Genesis keys
-    , PParams         -- Initial protocol parameters
+    , UTxO TxId
+    , BlockCount
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/Chain.hs]: Done!

File: byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/SigCnt.hs [hunk 25]
```diff
@@ -43,18 +37,13 @@ instance STS SIGCNT where
   transitionRules =
     [ do
         TRC ((pps, dms), sgs, vk) <- judgmentContext
-        let k = pps ^. stableAfter . to unBlockCount . to fromIntegral
-            t = pps ^. bkSgnCntT
+        let k  = pps ^. stableAfter . to unBlockCount . to fromIntegral
```
Comment (dnadales) [inline byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/SigCnt.hs]: Understandable it seems there is mismatch between the latest (?) version of this rule, and the implementation:

image

in particular `k` comes from the protocol parameters. Does this mean that we will model the global constants in this way? @mdimjasevic @nc6 any thoughts?

Comment (mdimjasevic) [reply byron/chain/executable-spec/src/Cardano/Spec/Chain/STS/Rule/SigCnt.hs]: I replied via chat (the Haskell implementation has been brought in line with the LaTeX version).

File: byron/chain/executable-spec/test/Main.hs [hunk 26]
```diff
@@ -14,7 +14,7 @@ main = defaultMain tests
   tests =
     testGroup "Chain"
     [ testGroup "Properties"
-      [ testProperty "Increasing slots" slotsIncrease
-      , testProperty "Block issuers are delegates" blockIssuersAreDelegates
+      [ -- testProperty "Increasing slots" slotsIncrease
```
Comment (dnadales) [inline byron/chain/executable-spec/test/Main.hs]: Why where these tests disabled?

Comment (mdimjasevic) [reply byron/chain/executable-spec/test/Main.hs]: I wasn't able to figure out how to get it working with new updated types.
